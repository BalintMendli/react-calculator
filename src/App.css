.App {
  text-align: center;
  background-color: #f5f5dc;
  height: 100vh;
  width: 100vw;
  display: flex;
  justify-content: center;
  align-items: center;
}

#calc-div{
  background-color: #000000;
  height: 420px;
  width: 300px;
  padding: 5px;
  box-sizing: border-box;
  display: grid;
  grid-template-columns: auto auto auto auto;
  grid-template-rows: auto auto auto auto auto auto;
  box-shadow: 2px 2px 10px 2px #81814a;
  border-radius: 5px;
}

#display{
  grid-column: 1 / 5;
  grid-row: 1;
  color: #ffffff;
  display: flex;
  align-items: flex-end;
  justify-content: flex-end;
  padding-bottom: 10px;
}

.buttons{
  color: #ffffff;
  display: flex;
  align-items: center;
  justify-content: center;
  border: 1px solid #000000;
  border-radius: 5px;
}

.buttons:hover{
  border: 1px solid #cccccc;
}

#AC{
  grid-column: 1 / 3;
  grid-row: 2;
  background-color: #AC3939;
}

#divide{
  grid-column: 3;
  grid-row: 2;
  background-color: #666666;
}

#multiply{
  grid-column: 4;
  grid-row: 2;
  background-color: #666666;
}

#seven{
  grid-column: 1;
  grid-row: 3;
  background-color: #4D4D4D;
}

#eight{
  grid-column: 2;
  grid-row: 3;
  background-color: #4D4D4D;
}

#nine{
  grid-column: 3;
  grid-row: 3;
  background-color: #4D4D4D;
}

#minus{
  grid-column: 4;
  grid-row: 3;
  background-color: #666666;
}

#four{
  grid-column: 1;
  grid-row: 4;
  background-color: #4D4D4D;
}

#five{
  grid-column: 2;
  grid-row: 4;
  background-color: #4D4D4D;
}

#six{
  grid-column: 3;
  grid-row: 4;
  background-color: #4D4D4D;
}

#plus{
  grid-column: 4;
  grid-row: 4;
  background-color: #666666;
}

#one{
  grid-column: 1;
  grid-row: 5;
  background-color: #4D4D4D;
}

#two{
  grid-column: 2;
  grid-row: 5;
  background-color: #4D4D4D;
}

#three{
  grid-column: 3;
  grid-row: 5;
  background-color: #4D4D4D;
}

#equal{
  grid-column: 4;
  grid-row: 5 / 7;
  background-color: #004466;
}

#zero{
  grid-column: 1 / 3;
  grid-row: 6;
  background-color: #4D4D4D;
}

#point{
  grid-column: 3;
  grid-row: 6;
  background-color: #4D4D4D;
}