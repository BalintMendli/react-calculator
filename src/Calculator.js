import React from 'react';
import PropTypes from 'prop-types';

const Calculator = ({
  display,
  handleClear,
  handleEval,
  handleNum,
  handleOp,
  handlePoint,
}) => (
  <div id="calc-div">
    <div id="display">{display}</div>
    <button type="button" id="clear" className="buttons" onClick={handleClear}>
      AC
    </button>
    <button type="button" id="divide" className="buttons" onClick={handleOp}>
      /
    </button>
    <button type="button" id="multiply" className="buttons" onClick={handleOp}>
      *
    </button>
    <button type="button" id="seven" className="buttons" onClick={handleNum}>
      7
    </button>
    <button type="button" id="eight" className="buttons" onClick={handleNum}>
      8
    </button>
    <button type="button" id="nine" className="buttons" onClick={handleNum}>
      9
    </button>
    <button type="button" id="subtract" className="buttons" onClick={handleOp}>
      -
    </button>
    <button type="button" id="four" className="buttons" onClick={handleNum}>
      4
    </button>
    <button type="button" id="five" className="buttons" onClick={handleNum}>
      5
    </button>
    <button type="button" id="six" className="buttons" onClick={handleNum}>
      6
    </button>
    <button type="button" id="add" className="buttons" onClick={handleOp}>
      +
    </button>
    <button type="button" id="one" className="buttons" onClick={handleNum}>
      1
    </button>
    <button type="button" id="two" className="buttons" onClick={handleNum}>
      2
    </button>
    <button type="button" id="three" className="buttons" onClick={handleNum}>
      3
    </button>
    <button type="button" id="equals" className="buttons" onClick={handleEval}>
      =
    </button>
    <button type="button" id="zero" className="buttons" onClick={handleNum}>
      0
    </button>
    <button
      type="button"
      id="decimal"
      className="buttons"
      onClick={handlePoint}
    >
      .
    </button>
  </div>
);

Calculator.propTypes = {
  display: PropTypes.string.isRequired,
  handleClear: PropTypes.func.isRequired,
  handleEval: PropTypes.func.isRequired,
  handleNum: PropTypes.func.isRequired,
  handleOp: PropTypes.func.isRequired,
  handlePoint: PropTypes.func.isRequired,
};

export default Calculator;
