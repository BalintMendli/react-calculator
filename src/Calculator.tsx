import React from 'react';

interface CalculatorProps {
  display: string;
  handleClear: React.MouseEventHandler<HTMLButtonElement>;
  handleEval: React.MouseEventHandler<HTMLButtonElement>;
  handleNum: React.MouseEventHandler<HTMLButtonElement>;
  handleOp: React.MouseEventHandler<HTMLButtonElement>;
  handlePoint: React.MouseEventHandler<HTMLButtonElement>;
}

const Calculator = ({
  display,
  handleClear,
  handleEval,
  handleNum,
  handleOp,
  handlePoint,
}: CalculatorProps): JSX.Element => (
  <div id="calc-div">
    <div id="display">{display}</div>
    <button type="button" id="clear" className="buttons" onClick={handleClear}>
      AC
    </button>
    <button type="button" id="divide" className="buttons" onClick={handleOp}>
      /
    </button>
    <button type="button" id="multiply" className="buttons" onClick={handleOp}>
      *
    </button>
    <button type="button" id="seven" className="buttons" onClick={handleNum}>
      7
    </button>
    <button type="button" id="eight" className="buttons" onClick={handleNum}>
      8
    </button>
    <button type="button" id="nine" className="buttons" onClick={handleNum}>
      9
    </button>
    <button type="button" id="subtract" className="buttons" onClick={handleOp}>
      -
    </button>
    <button type="button" id="four" className="buttons" onClick={handleNum}>
      4
    </button>
    <button type="button" id="five" className="buttons" onClick={handleNum}>
      5
    </button>
    <button type="button" id="six" className="buttons" onClick={handleNum}>
      6
    </button>
    <button type="button" id="add" className="buttons" onClick={handleOp}>
      +
    </button>
    <button type="button" id="one" className="buttons" onClick={handleNum}>
      1
    </button>
    <button type="button" id="two" className="buttons" onClick={handleNum}>
      2
    </button>
    <button type="button" id="three" className="buttons" onClick={handleNum}>
      3
    </button>
    <button type="button" id="equals" className="buttons" onClick={handleEval}>
      =
    </button>
    <button type="button" id="zero" className="buttons" onClick={handleNum}>
      0
    </button>
    <button
      type="button"
      id="decimal"
      className="buttons"
      onClick={handlePoint}
    >
      .
    </button>
  </div>
);

export default Calculator;
