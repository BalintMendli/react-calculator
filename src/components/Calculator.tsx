import React from 'react';
import styled from 'styled-components/macro';
import { Label } from 'constants/buttons';
import { heightPx, widthPx } from 'constants/dimensions';
import Display from './Display';
import BtnPanel from './BtnPanel';

interface CalculatorProps {
  result: string;
  expression: string;
  error: string;
  handleClick: (label: Label) => void;
}

interface StyledProps {
  widthPx: number;
  heightPx: number;
}

const StyledCalculator = styled.div<StyledProps>`
  height: ${(props): string => props.heightPx.toString()}px;
  width: ${(props): string => props.widthPx.toString()}px;
  background-color: #0b0c11;
  padding: 5px;
  display: flex;
  flex-direction: column;
  box-shadow: 1px 2px 10px rgba(0, 0, 0, 0.5);
  border-radius: 5px;
`;

const Calculator = ({
  result,
  expression,
  error,
  handleClick,
}: CalculatorProps): JSX.Element => (
  <StyledCalculator id="calculator" widthPx={widthPx} heightPx={heightPx}>
    <Display result={result} expression={expression} error={error} />
    <BtnPanel handleClick={handleClick} />
  </StyledCalculator>
);

export default Calculator;
