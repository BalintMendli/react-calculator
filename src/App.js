import React, { Component } from 'react';
import './App.css';

class App extends Component {
  render() {
    return (
      <div className="App">
        <div id="calc-div">
          <div id="display">00000</div>
          <div id="AC" className="buttons">AC</div>
          <div id="divide" className="buttons">/</div>
          <div id="multiply" className="buttons">x</div>
          <div id="seven" className="buttons">7</div>
          <div id="eight" className="buttons">8</div>
          <div id="nine" className="buttons">9</div>
          <div id="minus" className="buttons">-</div>
          <div id="four" className="buttons">4</div>
          <div id="five" className="buttons">5</div>
          <div id="six" className="buttons">6</div>
          <div id="plus" className="buttons">+</div>
          <div id="one" className="buttons">1</div>
          <div id="two" className="buttons">2</div>
          <div id="three" className="buttons">3</div>
          <div id="equal" className="buttons">=</div>
          <div id="zero" className="buttons">0</div>
          <div id="point" className="buttons">.</div>
        </div>
      </div>
    );
  }
}

export default App;
