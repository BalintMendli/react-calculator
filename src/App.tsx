import React, { Component } from 'react';
import './App.css';
import { evaluate } from 'mathjs';
import Calculator from './Calculator';

interface AppState {
  displayValue: string;
  isResult: boolean;
}

interface EventObj {
  target: { innerHTML: string };
}

class App extends Component<{}, AppState> {
  private appDiv = React.createRef<HTMLDivElement>();

  constructor(props: {}) {
    super(props);
    this.state = {
      displayValue: '0',
      isResult: false
    };
    this.handleEval = this.handleEval.bind(this);
    this.handleNum = this.handleNum.bind(this);
    this.handleClear = this.handleClear.bind(this);
    this.handleOp = this.handleOp.bind(this);
    this.handlePoint = this.handlePoint.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
  }

  componentDidMount() {
    this.appDiv.current!.focus();
  }

  handleEval() {
    const { displayValue } = this.state;
    const lastOp =
      ['/', '*', '+', '-'].indexOf(displayValue[displayValue.length - 1]) !==
      -1;
    const lastPoint = displayValue.slice(-1) === '.';
    if (!lastOp && !lastPoint) {
      this.setState(prevState => ({
        displayValue: (
          Math.round(1000000000000 * evaluate(prevState.displayValue)) /
          1000000000000
        ).toString(),
        isResult: true
      }));
    }
  }

  handleNum(e: React.MouseEvent<HTMLButtonElement, MouseEvent> | EventObj) {
    const { displayValue, isResult } = this.state;
    const target = e.target as HTMLElement;
    const value = target.innerHTML;
    const lastZero = displayValue.slice(-1) === '0';
    const isMaxDigit = displayValue.length >= 18;
    if (displayValue === '0' || isResult) {
      this.setState({
        displayValue: value,
        isResult: false
      });
    } else if (!isMaxDigit && !lastZero) {
      this.setState(prevState => ({
        displayValue: prevState.displayValue + value
      }));
    }
  }

  handleOp(e: React.MouseEvent<HTMLButtonElement, MouseEvent> | EventObj) {
    const { displayValue } = this.state;
    const target = e.target as HTMLElement;
    const value = target.innerHTML;
    const lastOp =
      ['/', '*', '+', '-'].indexOf(displayValue[displayValue.length - 1]) !==
      -1;
    const lastPoint = displayValue.slice(-1) === '.';
    const isMaxDigit = displayValue.length >= 18;
    if (displayValue === '0' && (value === '+' || value === '-')) {
      this.setState({
        displayValue: value,
        isResult: false
      });
    } else if (!lastOp && !lastPoint && displayValue !== '0' && !isMaxDigit) {
      this.setState(prevState => ({
        displayValue: prevState.displayValue + value,
        isResult: false
      }));
    } else if (lastOp) {
      this.setState(prevState => ({
        displayValue: prevState.displayValue.slice(0, -1) + value,
        isResult: false
      }));
    }
  }

  handlePoint() {
    const { displayValue, isResult } = this.state;
    const displayArr = displayValue.split(/[+\-*/]/);
    const pointCond = displayArr[displayArr.length - 1].indexOf('.') !== -1;
    const isMaxDigit = displayValue.length >= 18;
    if (!pointCond && !isResult && !isMaxDigit) {
      this.setState(prevState => ({
        displayValue: `${prevState.displayValue}.`
      }));
    } else if (isResult) {
      this.setState({
        displayValue: '0.',
        isResult: false
      });
    }
  }

  handleClear() {
    this.setState({ displayValue: '0' });
  }

  handleKeyDown(e: React.KeyboardEvent<HTMLDivElement>) {
    if (e.key === 'Enter') {
      e.preventDefault();
      this.handleEval();
    }
    if (e.key === '.' || e.key === ',' || e.key === 'Decimal') {
      this.handlePoint();
    }
    if (
      ['/', '*', '+', '-', 'Add', 'Subtract', 'Multiply', 'Divide'].indexOf(
        e.key
      ) !== -1
    ) {
      let char = e.key;
      if (e.key === 'Add') {
        char = '+';
      }
      if (e.key === 'Subtract') {
        char = '-';
      }
      if (e.key === 'Multiply') {
        char = '*';
      }
      if (e.key === 'Divide') {
        char = '/';
      }
      this.handleOp({ target: { innerHTML: char } });
    }
    if (
      ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'].indexOf(e.key) !== -1
    ) {
      this.handleNum({ target: { innerHTML: e.key } });
    }
    if (e.key === 'Delete' || e.key === 'Backspace') {
      this.handleClear();
    }
  }

  render() {
    const { displayValue } = this.state;
    return (
      <div
        className="App"
        role="tabpanel"
        tabIndex={0}
        ref={this.appDiv}
        onKeyDown={this.handleKeyDown}
      >
        <Calculator
          handleClear={this.handleClear}
          handleOp={this.handleOp}
          handleNum={this.handleNum}
          handleEval={this.handleEval}
          handlePoint={this.handlePoint}
          display={displayValue}
        />
      </div>
    );
  }
}

export default App;
