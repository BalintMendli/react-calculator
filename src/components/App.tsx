import React, { Component } from 'react';
import styled from 'styled-components/macro';
import { getLabelFromKey } from 'logic/utils';
import handleInput from 'logic/handleInput';
import { Label } from 'constants/buttons';
import Calculator from './Calculator';
import { AppState } from './AppState';

const StyledApp = styled.main`
  text-align: center;
  background-color: #a6cfe2;
  height: 100vh;
  width: 100vw;
  display: flex;
  justify-content: center;
  align-items: center;
`;

class App extends Component<{}, AppState> {
  private appDiv = React.createRef<HTMLDivElement>();

  constructor(props: {}) {
    super(props);
    this.state = {
      displayValue: '0',
      expression: '',
      isResult: false,
      error: '',
    };
    this.handleClick = this.handleClick.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
  }

  componentDidMount(): void {
    if (this.appDiv.current) this.appDiv.current.focus();
  }

  handleClick(label: Label): void {
    this.setState(
      prevState =>
        handleInput(label, { ...prevState }) as Pick<AppState, keyof AppState>,
    );
  }

  handleKeyDown(e: React.KeyboardEvent<HTMLDivElement>): void {
    e.preventDefault();
    const label = getLabelFromKey(e.key);
    if (label)
      this.setState(
        prevState =>
          handleInput(label, prevState) as Pick<AppState, keyof AppState>,
      );
  }

  render(): JSX.Element {
    const { displayValue, expression, error } = this.state;
    return (
      <StyledApp
        id="main"
        tabIndex={0}
        ref={this.appDiv}
        onKeyDown={this.handleKeyDown}
      >
        <Calculator
          handleClick={this.handleClick}
          result={displayValue}
          expression={expression}
          error={error}
        />
      </StyledApp>
    );
  }
}

export default App;
