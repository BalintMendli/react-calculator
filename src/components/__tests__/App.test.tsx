import React from 'react';
import App from 'components/App';
import Calculator from 'components/Calculator';
import { shallow, ShallowWrapper } from 'enzyme';

describe('App component', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  it('renders correctly', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('renders main div', () => {
    expect(wrapper.find('#main').length).toEqual(1);
  });

  it('renders Calculator', () => {
    expect(wrapper.find(Calculator).length).toEqual(1);
  });
});
