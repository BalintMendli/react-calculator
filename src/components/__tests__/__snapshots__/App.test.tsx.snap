// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders correctly 1`] = `
"<App__StyledApp id=\\"main\\" tabIndex={0} onKeyDown={[Function: bound handleKeyDown]}>
  <Calculator handleClick={[Function: bound handleClick]} result=\\"0\\" expression=\\"\\" error=\\"\\" />
</App__StyledApp>"
`;
